<a href="https://colab.research.google.com/github/Adams-ke/PYTHON/blob/main/BERT_intro_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### BERT Embeddings for Log Classification
####
#### This notebook demonstrates how to generate embeddings using a pre-trained BERT model
#### for log message classification. BERT (Bidirectional Encoder Representations from Transformers)
#### converts text logs into dense numerical vectors (embeddings) that capture semantic meaning.
####
#### These embeddings serve as input features for a downstream classifier to automatically
#### categorize log messages (e.g., security_breach, hardware_failure, system_error).
####
#### Key concepts covered:
#### - Tokenization: Converting raw text to token IDs with [CLS] and [SEP] special tokens
#### - Padding & Truncation: Handling variable-length inputs to a fixed max length
#### - Input IDs: Mapping tokens to BERT's vocabulary indices
#### - Attention Masks: Distinguishing real tokens from padding
#### - Token Type IDs: Differentiating between multiple sentences
#### - Pooled Output: Sentence-level embeddings for classification
#### - Sequence Output: Token-level embeddings for each position
####
#### The final pooled_output vectors (shape: batch_size, 768) are used as features
#### to train a classifier for automated log analysis.


In [ ]:

import tensorflow_hub as hub
import tensorflow_text as text

In [ ]:
import kagglehub

# Download latest version
path = kagglehub.model_download("tensorflow/bert/tensorFlow2/bert-en-uncased-l-8-h-768-a-12")

print("Path to model files:", path)


100%|██████████| 2.67k/2.67k [00:00<00:00, 4.46MB/s]



  0%|          | 0.00/226k [00:00<?, ?B/s]



  0%|          | 0.00/131k [00:00<?, ?B/s]




  0%|          | 0.00/6.29M [00:00<?, ?B/s]





  0%|          | 0.00/312M [00:00<?, ?B/s]

100%|██████████| 131k/131k [00:00<00:00, 786kB/s]

100%|██████████| 226k/226k [00:00<00:00, 1.00MB/s]



 16%|█▌        | 1.00M/6.29M [00:00<00:02, 2.57MB/s]



  0%|          | 1.00M/312M [00:00<02:12, 2.47MB/s]


100%|██████████| 6.29M/6.29M [00:00<00:00, 11.0MB/s]




  1%|          | 3.00M/312M [00:00<00:47, 6.76MB/s]



  3%|▎         | 8.00M/312M [00:00<00:17, 17.9MB/s]



  4%|▍         | 13.0M/312M [00:00<00:11, 26.8MB/s]



  6%|▌         | 18.0M/312M [00:00<00:09, 32.9MB/s]



  8%|▊         | 24.0M/312M [00:00<00:07, 40.7MB/s]



  9%|▉         | 29.0M/312M [00:01<00:06, 42.8MB/s]



 11%|█         | 35.0M/312M [00:01<00:07, 36.7MB/s]



 13%|█▎        | 41.0M/312M [00:01<00:07, 36.8MB/s]



 14%|█▍        | 45.0M/312M [00:01<00:08, 31.5MB/s]



 16%|█▌        | 50.0M/312M [00:01<00:08, 33.4MB/s]



 19%|█▊        | 58.0M/312M [00:01<00:06, 40.3MB/s]



 20%|██        | 63.0M/312M [00:02<00:06, 41.4MB/s]



 22%|██▏       | 70.

Path to model files: /root/.cache/kagglehub/models/tensorflow/bert/tensorFlow2/bert-en-uncased-l-8-h-768-a-12/2


In [ ]:
import kagglehub

# Download latest version
preprocess_path = kagglehub.model_download("tensorflow/bert/tensorFlow2/en-uncased-preprocess")

print("Path to model files:", path)



100%|██████████| 4.55k/4.55k [00:00<00:00, 8.14MB/s]



100%|██████████| 1.75k/1.75k [00:00<00:00, 1.20MB/s]



100%|██████████| 276/276 [00:00<00:00, 506kB/s]


  0%|          | 0.00/226k [00:00<?, ?B/s]

  0%|          | 0.00/1.74M [00:00<?, ?B/s]
100%|██████████| 226k/226k [00:00<00:00, 1.02MB/s]


100%|██████████| 1.74M/1.74M [00:00<00:00, 4.13MB/s]

Path to model files: /root/.cache/kagglehub/models/tensorflow/bert/tensorFlow2/bert-en-uncased-l-8-h-768-a-12/2


In [ ]:
bert_preprocess_model = hub.KerasLayer(preprocess_path)

In [ ]:
text_test = ['nice movie indeed', 'I love python programming']
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_type_ids', 'input_mask', 'input_word_ids'])

In [ ]:
text_preprocessed['input_mask']


<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

## Why 5 tokens instead of 3?
###The original sentence has 3 words, but BERT's tokenizer adds:
###1. [CLS] - Special token added to the beginning
###   2. [SEP] - Separator token added to the end
### This increases the token count from 3 to 5.

### The shape shows 218 tokens - this is the maximum sequence length.
### Shorter sentences are padded with [PAD] tokens to reach this length.
### Only the first few tokens contain actual words; the rest are padding.

In [ ]:
text_preprocessed['input_type_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]],
      dtype=int32)>

#### input_type_ids indicate which sentence each token belongs to.
#### This is useful for tasks involving multiple sentences (e.g.,
#### next sentence prediction, question answering, or duplicate detection).

In [ ]:
text_preprocessed['input_word_ids']

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[  101,  3835,  3185,  5262,   102,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0,     0,     0,
            0,     0,     0, 

#### input_word_ids: Maps each token in the input to a unique integer ID
#### from BERT's pre-trained vocabulary. For example:
#### "hello world" -> [7592, 2088] (token IDs)
#### Special tokens like [CLS] and [SEP] also get their own IDs.

In [10]:
bert_model = hub.KerasLayer(path)

bert_results = bert_model(text_preprocessed)
bert_results.keys()

dict_keys(['pooled_output', 'sequence_output', 'encoder_outputs', 'default'])

In [11]:
bert_results['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[ 0.1897155 , -0.80589736, -0.2773873 , ...,  0.45073768,
        -0.10450462, -0.96556735],
       [-0.22860238,  0.9856232 , -0.33071098, ...,  0.2098506 ,
        -0.06034147,  0.28747988]], dtype=float32)>

#### pooled_output: Shape (2, 768)
####  - Row 1: [0.1897, -0.8059, -0.2774, ..., 0.4507, -0.1045, -0.9656] for the first sentence
####   - Row 2: [-0.2286, 0.9856, -0.3307, ..., 0.2099, -0.0603, 0.2875] fro the 2nd sentence
#### Each row is a fixed-length embedding representing a complete sentence.
#### The classifier model will use these 768 features as input.

In [12]:
bert_results['sequence_output'] #Individual word embedding vectors

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.0036357 , -0.28434426, -0.47290474, ...,  0.325473  ,
         -0.41943797, -0.62765646],
        [-0.0323559 ,  0.06783202, -0.05992878, ...,  0.41025168,
         -0.43100667, -0.2712654 ],
        [-0.03054966, -0.04882111, -0.47103468, ...,  0.04525398,
         -0.6152221 , -0.01579081],
        ...,
        [ 0.24068798, -0.23280784,  0.10993624, ...,  0.5194389 ,
         -0.32861653, -0.08736786],
        [ 0.36942244, -0.2151529 ,  0.224934  , ...,  0.53628105,
         -0.5569312 , -0.16120777],
        [ 0.14861275, -0.18665476,  0.02004697, ...,  0.04710242,
         -0.36566696, -0.02494475]],

       [[ 0.19766223, -0.9127148 , -0.7452311 , ...,  0.14361769,
          0.42593843, -0.13625759],
        [-0.22837275, -0.07495797, -0.46998158, ...,  0.5539796 ,
         -0.03479493,  0.5935309 ],
        [-0.35663778,  0.10392379, -0.10257839, ..., -0.17184874,
         -0.06619662,  0.41135925],
        ...,


#### Each sentence is represented by a 768-dimensional vector.
#### Even padding tokens (the 0s added to fill to max_length like 128) get embeddings,
#### since BERT is contextualized and computes representations for every token.
#### The [CLS] token aggregates information from the entire sequence,
#### and pooled_output is derived from this [CLS] token's final state.

In [15]:
len(bert_results['encoder_outputs'])# this corresponds to the model(bert-en-uncased-l-8-h-768-a-12) that was used.

8

In [16]:
bert_results['encoder_outputs'][0]# the first one

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[-0.01882518, -0.29990554,  0.04615699, ..., -0.25180873,
          0.04540905, -0.502308  ],
        [ 0.28459007, -0.5706699 , -0.16023248, ...,  0.4317318 ,
         -0.34586465, -0.71410626],
        [-1.6680412 , -0.8349872 , -0.39382562, ...,  0.76850665,
         -0.91023403, -0.1404809 ],
        ...,
        [ 0.1994154 ,  0.3246878 ,  0.06833257, ...,  0.8716121 ,
          0.11859913,  0.28596997],
        [ 0.03121053,  0.4085598 ,  0.17950541, ...,  0.85446864,
          0.15439412,  0.17374964],
        [ 0.13004237,  0.43972987,  0.34364176, ...,  0.80595267,
          0.12428077,  0.20682628]],

       [[-0.01700499, -0.19240972, -0.08877143, ..., -0.20897725,
          0.00780325, -0.52373415],
        [ 0.9613436 , -0.54499024,  0.57278144, ...,  0.31552413,
         -0.5348336 ,  0.61667746],
        [ 0.34910297, -0.49218488, -0.8142506 , ...,  0.29429176,
          0.3501438 ,  0.5974784 ],
        ...,

In [21]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']



<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

#### Convenience and clarity:
#### - sequence_output: Explicit, obvious name for token-level embeddings
#### - encoder_outputs[-1]: Less obvious, requires knowing indexing
####
#### They keep both for:
#### 1. Backward compatibility (older code uses sequence_output)
#### 2. Flexibility (encoder_outputs lets you access intermediate layers)